Issue #3: add SSL support

cjrh opened the pull request:
I went through several designs, but this is by far the cleanest: let the user build their own `SSLContext` and pass that in as part of `StompConfig`.

I tested this against **ActiveMQ 5.13.0**, running under docker.  It was a real pain to set up an amq instance with SSL enabled.  (The one I used is `cjrh/amq` on dockerhub).

I tested with **Python 3.6.0**, although it should "just work" under 2.7. I just haven't run the tests under 2.7 yet.

- For some reason _Twisted_ complained about _pyopenssl_  and _service_identity_ being missing, but I already had _Twisted_ installed and it didn't install them as deps. I didn't look into it any deeper than that, so you may want to verify whether they really do need to be added to the `install_requires` or not.
- As for testing, it was easiest to simply rerun all the existing integration tests, but under SSL, rather than come up with new tests. I don't know which direction you wanted to go in here.

cjrh commented:
I presume that you'd like me to update the documentation too; I'll do that once it looks like this PR is likely to be accepted.

nikipore commented:
Thanks @cjrh! I skimmed over the code and have the following remarks to begin with:

* `stompest` convention for attributes and local variables is `camelCase`.
* Is there a way to get along without the new dependencies (viz. to inject SSL)?

I'll look into it more deeply and test the code when I find some time (could take a while, I'm rather busy).

cjrh commented:
Thanks for you comments.  I hope to get to them today.

The Twisted documentation says that TLS support requires pyopenssl:

https://twistedmatrix.com/documents/current/core/howto/ssl.html

> Using TLS in Twisted requires that you have pyopenssl installed. A quick test to verify that you do is to run from OpenSSL import SSL at a python prompt and not get an error.

I think it will be fine to remove pyopenssl as a stompest dependency, and instead require the user to add it if they use the SSL feature under async.  A note about this will be added to the documentation.  It will probably be necessary to add pyopenssl under `test_requires` though, else the SSL tests would fail.

I'll look into removing the `service_identity` dependency too, and fix the camelCase.

cjrh commented:
Ok, `service_identity` was mentioned also in the Twisted documentation, here:

https://twistedmatrix.com/documents/current/core/howto/ssl.html#connecting-to-public-servers

> Note
>
>To properly validate your hostname parameter according to RFC6125, please also install the “service_identity” and “idna” packages from PyPI. Without this package, Twisted will currently make a conservative guess as to the correctness of the server’s certificate, but this will reject a large number of potentially valid certificates. service_identity implements the standard correctly and it will be a required dependency for TLS in a future release of Twisted.

I'll add a link in the documentation to this section.

cjrh commented:
Hi @nikipore 

- I've changed the variable naming to your convention (mostly `ssl_context` became `sslContext` everywhere).
- I've added an SSL integration test for the sync client
- I've removed _pyopenssl_ and _service_identity_ from `install_requires` and into `tests_require`, but only for the async `setup.py`.  _pyopenssl_ is definitely required for TLS support in Twisted.

On this last point, I am unsure about the value of not including them by default.  It means that SSL support will be unusable until pyopenssl (for the async client) is installed, and the only way to know that you need it is to either read the documentation upfront, or read the error messages that will occur if _pyopenssl_ is missing.

Another way to go is to make SSL dependencies an `extras_require`, so for example if the user wants SSL support, they would install stompest like this:

```
$ pip install stompest.async[ssl]
```

That way, the SSL dependencies are only installed if specified.  (and of course, they'd have to then also remain in `tests_require` since they're always needed for the async tests.

Anyway, it's your project so you are the king; I will make it however you want :)

If it looks good, I can begin making adding documentation.  Hints about which pages to affect, sections etc., would be greatly appreciated since it would save me time.

By the way, thank you for all your hard work on maintaining stompest.  Open source maintainers never get enough thanks!

cjrh commented:
Oh, and I forgot to mention that I'll squash the commits before merge too (although github nowadays can do that do).

nikipore commented:
Thank you for your great work and kind words. Python 3 and SSL support have always been the two missing features which troubled me most, so I'm very glad that you walked this extra mile.

* I'd opt for the `extras_require` approach, but delegating to Twisted's `tls` extra, viz. `extras_require={'tls': 'twisted[tls]'}`. Hopefully that works as expected, cf. [this pip issue](https://github.com/pypa/pip/issues/3189). And I think TLS or TLS/SSL is the way to call SSL these days.
* It looks great now, so you're good to move on to documentation. I'd suggest either adding "TLS/SSL" subsections in both `sync.rst` and `async.rst` (if you want to stress the differences between the two) or a new section `tls.rst` (if you want to stress the commonalities). Please add installation instructions (including the ActiveMQ settings), a minimal working example for each client, and link to the relevant docs for further reading, e.g. [Twisted – Using TLS](https://twistedmatrix.com/documents/current/core/howto/ssl.html), [Python 3 – TLS/SSL wrapper for socket objects](https://docs.python.org/3/library/ssl.html), or [Apache Active MQ – Stomp](http://activemq.apache.org/stomp.html).
* Since you'll have to set up the documentation build process anyway, you would do me a great favor if you could also tweak the documentation Makefile and/or add a "documentation" section s.t. the gentle reader will be enabled to build the documentation from scratch.

cjrh commented:
Excellent feedback, thank you.  I'll proceed exactly as you have laid out.

cjrh commented:
I've made all the changes you've set out.

The plan to use `extras_require={'tls': 'twisted[tls]'}` was very smart, but unfortunately still doesn't work upstream.  There are open bug reports still for this feature:

- https://github.com/pypa/pip/issues/3516
- https://github.com/pypa/pip/pull/3878

For this reason, I've change the `install_requires` to specify `twisted[tls] >= 16.4.0` directly.  I know you wanted to keep the SSL-related dependencies separate if possible, but it doesn't seem like there is an easy answer here.  Suggestions?

For the documentation, I've added instructions on how to build it in the `README.markdown` file.  I've also changed the target build directory (the HTML output) to be inside the `doc/` directory, and added `doc/stompest-doc` to the `.gitignore` list. It felt more natural to me to have the documentation build output appear where `$ make html` is run.  I also deleted the `make` executable script, since it seemed tied to a specific environment configuration that required virtualenvwrapper, and other environment variables.  In further support of the documentation, I added an `extras_require={'doc': 'sphinx'}` so that sphinx can be installed this way.  This is explained in the README.  

Everything else, such as the changes in dates, and CHANGES.txt and version numbers and so on have been done as you described.

nikipore commented:
I relaxed the dependency from `twisted[tls]` to `twisted` and compensated for it by installation instructions in the readme/documentation. Release 2.3.0 is out of the door. Thanks again!
